CMakeLists.txt

```diff
@@ -25,6 +25,10 @@ endif()
 
 list(APPEND CMAKE_MODULE_PATH "${CMAKE_SOURCE_DIR}/cmake/modules/")
 
+if(${CMAKE_BUILD_TYPE} MATCHES "Debug")
+  add_definitions(-DDEBUG)
```

Comment:
@tchaikov without my change, and use gdb to debug, and input "next", the cursor should go to next code line, but it usually go to up line...

There is a comparison on build/src/CMakeFiles/ceph-mon.dir/flags.make:
No add_definitions(-DDEBUG):
CXX_FLAGS =  -Wall -Wtype-limits -Wignored-qualifiers -Winit-self -Wpointer-arith -Werror=format-security -fno-strict-aliasing -fsigned-char -rdynamic -Wno-unknown-pragmas -Wp,-U_FORTIFY_SOURCE,-D_FORTIFY_SOURCE=2 -Wp,-U_FORTIFY_SOURCE,-D_FORTIFY_SOURCE=2 -std=c++11 -I/usr/include/nss -I/usr/include/nspr -fno-builtin-malloc -fno-builtin-calloc -fno-builtin-realloc -fno-builtin-free **-O2 -g**  -fPIE -isystem /root/ceph_commit/ceph/build/boost/include -I/root/ceph_commit/ceph/build/src/include -I/root/ceph_commit/ceph/src -isystem /root/ceph_commit/ceph/build/include -I/root/ceph_commit/ceph/src/xxHash -I/root/ceph_commit/ceph/src/rocksdb/include    -DHAVE_CONFIG_H -D__CEPH__ -D_REENTRANT -D_THREAD_SAFE -D__STDC_FORMAT_MACROS -D_GNU_SOURCE -D_FILE_OFFSET_BITS=64

With add_definitions(-DDEBUG):
CXX_FLAGS =  -Wall -Wtype-limits -Wignored-qualifiers -Winit-self -Wpointer-arith -Werror=format-security -fno-strict-aliasing -fsigned-char -rdynamic -Wno-unknown-pragmas  -ftemplate-depth-1024 -Wno-invalid-offsetof -Wnon-virtual-dtor -Wstrict-null-sentinel -Woverloaded-virtual -std=c++11 -I/usr/include/nss -I/usr/include/nspr -fno-builtin-malloc -fno-builtin-calloc -fno-builtin-realloc -fno-builtin-free **-g** -fPIE -isystem /root/ceph_commit/ceph/build/boost/include -I/root/ceph_commit/ceph/build/src/include -I/root/ceph_commit/ceph/src -isystem /root/ceph_commit/ceph/build/include -I/root/ceph_commit/ceph/src/xxHash -I/root/ceph_commit/ceph/src/rocksdb/include    -DHAVE_CONFIG_H -D__CEPH__ -D_REENTRANT -D_THREAD_SAFE -D__STDC_FORMAT_MACROS -D_GNU_SOURCE -D_FILE_OFFSET_BITS=64

So after my change, the optimization is gone, and gdb works well.